### PR TITLE
Viewing sky/weather over player position

### PR DESCRIPTION
**Pull request by int-ua**
#### Summary
```SUMMARY: Features "Ability to view weather around player position"```

#### Purpose of change
Give player a chance to see sky/weather overhead for trying to predict weather changes and act accordingly.

#### Describe the solution
Added new action "Look at the sky" that displays weather map limited to line of sight with double visibility. Currently it's under "Look..." in actions menu (`Enter`).
Overmap weather glyphs and colors changed (see screenshot below).
Debug weather and this view is limited to z 10 only.

#### Describe alternatives you've considered
Describing weather overhead by text. Requires a lot of code and text to cover a wide area.

#### Additional context
Along with the benefit in itself this feature will allow adding usable and realistic ability to sun-dry some food.
Requesting comments on:
- Do you see any problems in CPP code or with the changes in general?
- ~~Any suggestions on how to add a legend?~~
- ~~Should weather representation be changed? On this map only or along with debug weather map?~~
- ~~Where is the best place to store `sight_points` for viewing sky? It's currently hardcoded, but it's just a PoC. Is 20 a good value?~~

Current state of viewing debug weather (types captioned after making the screenshot):
third_weather_colors_

and player-visible part (usual overmap sight points doubled)
screenshot_20190130_202716

### CODE REVIEW

**Comment by davidpwbrown** [comment]
I really need to add debug menu stuff for wind, once I get my head around the debug options.
Currently, the weather overlay dosn't show any wind stuff, it really should.
having a dedicated "weather view" may mean some weather elements could be removed from UI , yeah.
But first I'd need to find a way to intertwine weather_type which has set types "sunny, cloudy, rainy" with the wind stuff.

**Comment by KITbutler** [comment]
Can that screen print special symbols, such as arrows representing wind direction? A regular human should be able to tell the wind on their own tile as well as get the direction (but very inaccurate speed) of any nearby clouds moving above. 

The weather reader bionic wouldn't realistically add much on this scale, the most I would expect is a [LIDAR/laser scattering system](https://en.wikipedia.org/wiki/Laser_guide_star) capable of detecting rainfall and the "exact" speed of clouds. Printing humidity, pressure, air temperature, depth-penetrating scans of storms etc. would require standing still for proper calibration followed by a lengthy analysis. If you even have the multi-wavelength tools to do it in the first place that is, a proper [weather radar](https://en.wikipedia.org/wiki/Weather_radar) is beyond our ability to carry *or* power.

No idea regarding the coding but the sidebar in your screenshot has plenty of space for all weather info we could conceivably have access to. If that is too hardcoded maybe you can make an overlay similar to the (f)ire or "look around"/mouse cursor panel?

**Comment by hyper2snyper** [comment]
You can make it so the higher survival skill you have the more specific the map gets. So at a survival skill of zero you can only tell dark clouds, and light ones apart. Whereas the higher your skill becomes the better you can differentiate what clouds do what.

**Comment by int-ua** [comment]
>  Temperature, humidity, pressure and wind.
But first three only with tools, of course.

>  intertwine weather_type which has set types "sunny, cloudy, rainy" with the wind stuff
Sounds relatively hard to do.
Did you also mean that wind direction should affect clouds movement? Since wind direction near ground not always represents wind direction near clouds it might be not necessary.

On showing wind direction. I kind of like the idea of an arrow with pluses even on this screen because player have a way to measure wind speed _directly_ only in one point. But maybe there is a better way that wasn't described yet. Cloud movement can be seen by waiting and looking at the sky again. Yes, in reality you can estimate cloud speed almost instantly if they have it above some threshold but I'm not sure if such thing as cloud movement direction is even present in current implementation. I just skimmed through `weather_generator::get_weather` and it looks like in place of time it's just using seed value, do I understand it correctly?

**Comment by int-ua** [comment]
On predicting weather in UI: that is a good idea, of course, but I'll limit this PR to viewing current state only. It's already hard for me as it is, sorry. Maybe someone else wants to add prediction?

File: src/handle_action.cpp [hunk 1]
```diff
@@ -1784,6 +1784,15 @@ bool game::handle_action()
                 ui::omap::display();
                 break;
 
+            case ACTION_SKY:
+                if ( m.is_outside( u.pos() ) ) {
```
Comment (Night-Pryanik) [inline src/handle_action.cpp]:
```suggestion
                if( m.is_outside( u.pos() ) ) {
```

File: src/handle_action.cpp [hunk 2]
```diff
@@ -1784,6 +1784,15 @@ bool game::handle_action()
...
+                    werase( w_terrain );
+                    ui::omap::display_visible_weather();
+                } else {
+                    add_msg( m_info, _( "You can't see the sky from here" ) );
```
Comment (Night-Pryanik) [inline src/handle_action.cpp]:
```suggestion
                    add_msg( m_info, _( "You can't see the sky from here." ) );
```

File: src/handle_action.cpp (hunk 1 above)
Comment (int-ua) [reply src/handle_action.cpp]: I'll try not to forget running `make astyle`, thank you. Will be added in the next commit.

File: src/handle_action.cpp (hunk 2 above)
Comment (int-ua) [reply src/handle_action.cpp]: Included in the next commit too.

**Comment by KITbutler** [comment]
> On showing wind direction. I kind of like the idea of an arrow with pluses even on this screen because player have a way to measure wind speed directly only in one point. But maybe there is a better way that wasn't described yet. Cloud movement can be seen by waiting and looking at the sky again. Yes, in reality you can estimate cloud speed almost instantly if they have it above some threshold but I'm not sure if such thing as cloud movement direction is even present in current implementation.

My idea was to display the local wind direction on cloud tiles, perhaps with a more stringent range limit than the overall map. The idea being that you can tell with your bare eyes which direction the clouds seem to be moving in at least once they reach a certain speed (i.e. low wind speeds would render as "stationary") even if you can't get a reasonable estimate of said speed. 

With a weather reader bionic or a portable device you could point it at a specific cloud tile (**only** clouds and rain/snowfall!) and use RADAR/LIDAR techniques to estimate the movement speed of said clouds, either as an "activity" with a time cost or "for free" as a sidebar display similar to how the overmap displays the location name of the currently selected tile.

> Since wind direction near ground not always represents wind direction near clouds it might be not necessary.

True, but do we have any underlying code which consistently models a two-layer atmosphere? The surface wind has to be powered by *something* and the rainclouds are also moved by wind so making it a single self-consistent layer would at least be a good first step for improvement otherwise. Adding RNG to readouts or expanding to a proper two-layer model can be done later if it is deemed worth the effort.

Full weather maps probably would require a network of actual weather stations, in particular the high/low-pressure regions commonly used to display wind directions would be almost impossible to construct from a single location. On the flip side simple fluid tube thermometers and barometers should be almost trivial to craft and can be calibrated using water (freezing+boiling for the thermometer, see Celsius scale) and some basic math.

**Comment by int-ua** [comment]
Changed some weather glyphs:

` ` both sunny and clear (kind of struggle to make a distinction visible, any ideas?)
`.` clouds without precipitation (cloudy)
`o` some precipitation (drizzle, flurries)
`8` precipitation (rain, snowing)
`%` storms

screenshot_20190129_111258

**Comment by int-ua** [comment]
> display the local wind direction on cloud tiles
Unfortunately, that's two pieces of information and we have only one symbol to display weather on one overmap tile. Can you please draw how you see it over my screenshot in the previous comment?

**Comment by int-ua** [comment]
> weather types themselves have various visibility modifiers
I've changed visibility to be just double of usual sight points. Is it ok?

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/wip-looking-at-the-sky-around-player-position/18594/1

**Comment by davidpwbrown** [comment]
> > Temperature, humidity, pressure and wind.
> 
> But first three only with tools, of course.
> 
> > intertwine weather_type which has set types "sunny, cloudy, rainy" with the wind stuff
> 
> Sounds relatively hard to do.
> Did you also mean that wind direction should affect clouds movement? Since wind direction near ground not always represents wind direction near clouds it might be not necessary.
> 
> On showing wind direction. I kind of like the idea of an arrow with pluses even on this screen because player have a way to measure wind speed _directly_ only in one point. But maybe there is a better way that wasn't described yet. Cloud movement can be seen by waiting and looking at the sky again. Yes, in reality you can estimate cloud speed almost instantly if they have it above some threshold but I'm not sure if such thing as cloud movement direction is even present in current implementation. I just skimmed through `weather_generator::get_weather` and it looks like in place of time it's just using seed value, do I understand it correctly?

Nah I just meant that there is already a nice neat weather_type that loads of other functions can call and the debug weather view can show ( as above ) , "SUNNY/CLOUDY/SNOW" etc but then wind is kinda tacked on as another thing, it'd just be neater and less confusing to have one place where you can ask the code "what is the weather here?" and get the conditions for wind and sun and rain and everything in one go.
  It would make this visualizer you are doing a bit simpler also.

**Comment by int-ua** [comment]
> get the conditions for wind and sun and rain and everything in one go.

So you want to add a new function that will return everything for a given overmap point? That sounds great. I've looked into `weather_datum.sight_penalty` usage and looks like it doesn't affect overmap visibility in current state. Should it?

**Comment by davidpwbrown** [comment]
As far as I can see it just effects player sight range for monsters and stuff - you can see on the minimap if you debug change weather, the visible circle shrinks.
But you could use it as a variable/guide to reference range in your map.

**Comment by Bidiguilo** [comment]
gosh my eye hurts after going full screen in that one photo lol

**Comment by int-ua** [comment]
> my eye hurts
You are not the first to say this and I'm starting to worry. Will changing colors be enough? Any other suggestions?

**Comment by int-ua** [comment]
`src/color.*` look horrible. Maybe that was unavoidable but it's a hell for trying to figure out how to put arbitrary color on both background and foreground.

**Comment by int-ua** [comment]
New colors:
second_weather_colors

**Comment by int-ua** [comment]
Should be easier to quickly distinguish weather with high DPI:
third_weather_colors_

**Comment by Achie72** [comment]
Why not use a color scheme that is used already, and helps to differentiate more easily?
for example: https://www.rainviewer.com/ has some pretty cool color scheme in hwich you can tell in a blink what's going on.

**Comment by Amneiger** [comment]
That does look good.  I don't see a legend; you might want to see if you can add one, to help people who aren't familiar with the map yet.

**Comment by int-ua** [comment]
>  i will now know when to roam around with my solar car when it is not going to rain all day

Just like in reality you can't be sure about how exactly weather is going to change. I've  run some tests and the part you see gives you only several hours estimation.

> thunderstorm is blending in with the rain too

Agreed, I'll try to improve that.

> Why not use a color scheme that is used already

Don't forget that this game is playable in text mode without tiles and color selection is extremely limited there. Check the Color manager in game. I've verified that this map is working as expected in Linux terminal but I've seen something about some Windows text mode having extreme shortage of colors - just 8. Can someone confirm that map is usable there please?

> I don't see a legend

Code in this PR is just an addition to the general overmap code. If this feature is accepted into the game it may be feasible moving this into a separate view and that would allow more freedom. But for now it's showing what weather is under the pointer just like with features on usual overmap view.

**Comment by int-ua** [comment]
New rain color:
screenshot_20190131_094814

**Comment by Achie72** [comment]
> Don't forget that this game is playable in text mode without tiles and color selection is extremely limited there. Check the Color manager in game. I've verified that this map is working as expected in Linux terminal but I've seen something about some Windows text mode having extreme shortage of colors - just 8. Can someone confirm that map is usable there please?

Sorry, i'm not really familiar with the code yet, just trying to start learning the code in a few days. 
This new colors look better imo. Still a bit hard to differentiate, but over time it's learnable. Definitely more clearer than the previous one.

**Comment by davidpwbrown** [comment]
this woudl work well with a revival of #27215

**Comment by kevingranade** [comment]
Summary suggestion:
SUMMARY: Features "Take a look to the sky just before you die"
: D

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/wip-looking-at-the-sky-around-player-position/18594/1

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/wip-looking-at-the-sky-around-player-position/18594/1

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/wip-looking-at-the-sky-around-player-position/18594/1

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/wip-looking-at-the-sky-around-player-position/18594/1

**Pull request merged**